src/coreclr/vm/jitinterface.cpp

```diff
@@ -9424,6 +9424,10 @@ CorInfoTypeWithMod CEEInfo::getArgType (
         }
         break;
 
+    case ELEMENT_TYPE_BYREF:
+        typeHnd = ptr.GetTypeHandleThrowing(pModule, &typeContext);
+        break;
```

Comment:
Can we do it for all types? If not, can we at least do it for `ELEMENT_TYPE_PTR` as well?

The documentation of the JIT-EE API should be updated (`vcTypeRet` should probably be renamed too). Also, what about crossgen2/NAOT implementations of this API?

Reply:
Does doing this cause the operand type of the byref/pointer to be loaded eagerly?

Reply:
> Does doing this cause the operand type of the byref/pointer to be loaded eagerly?

How do I check that? I presume since it's a parameter it's going to be loaded anyway?

Reply:
> Can we do it for all types? 

Any particular reason to do that? Isn't, like you mentioned, it might lead to redundant type load events we don't need?

Reply:
> How do I check that? I presume since it's a parameter it's going to be loaded anyway?

Maybe @jkotas can answer this question?

> Any particular reason to do that? Isn't, like you mentioned, it might lead to redundant type load events we don't need?

The reason would be to make the JIT-EE API more regular in its contract, and avoid us having to call back to the EE side in the other cases too (e.g. we want this information for class types as well, and we currently do another JIT-EE call to get it). We could remove `getArgClass` entirely if `getArgType` just returned the class handle always.

Reply:
> it might lead to redundant type load events we don't need?

Once we load the type, the future attempts to load the type are cheap. There is not such a thing as "redundant type load events".

The point is to avoid loading the type in the first place. As you can see in the `ELEMENT_TYPE_PTR` comment below, eagerly loading types that do not need to be loaded can even break stuff. It is 

I think the JIT should never need to load the unmanaged pointer types. It should treat all unmanaged pointers as opaque `void*`.

This optimization should be based on the type provided in the stfld metadata token.

Reply:
> This optimization should be based on the type provided in the stfld metadata token.

Actually, this might simplify changes a lot, let me try

Reply:
> This optimization should be based on the type provided in the stfld metadata token.

This would also be much more in line with how our normal modelling of aliasing and stores/loads work and effectively would mean that we aren't changing the aliasing model, so if we can do that it seems like it would be much cleaner overall.